msaad_test.go

```diff
@@ -0,0 +1,559 @@
+package authaus
+
+import (
+	"github.com/IMQS/log"
+	"github.com/stretchr/testify/assert"
+	"testing"
+	"time"
+)
+
+/** Test cases for msaad.go **/
+
+/*
+NOTICE : When adding tests here for new functions related to user credentials,
+make sure to check if similar functions in oauth.go does not need to be updated.
+Since OAUTH and MSAAD are similar in many ways, and both can directly affect
+user details in the system, slight difference in how they convert user details
+from the identity authority can cause issues. One example would be
+how the user's "Username" is treated.
+*/
+
+var testProvider *dummyMSAADProvider
+
+func getCentralMSAAD(t *testing.T) *Central {
+	var userStore UserStore
+	var sessionDB SessionDB
+	var permitDB PermitDB
+	var roleDB RoleGroupDB
+	var msaad MSAAD
+	msaad.SetConfig(ConfigMSAAD{
+		Verbose:              true,
+		DryRun:               false,
+		TenantID:             irrelevantUUID,
+		ClientID:             irrelevantUUID,
+		ClientSecret:         "abcdef",
+		MergeIntervalSeconds: 30,
+		DefaultRoles:         []string{"enabled"},
+		RoleToGroup: map[string]string{
+			"AZ_ROLE_1": "IMQS Group 1",
+			"AZ_ROLE_2": "IMQS Group 2",
+			"AZ_ROLE_3": "IMQS Group 3",
+		},
+		AllowArchiveUser:     true,
+		PassthroughClientIDs: []string{irrelevantUUID},
+	})
+	testProvider = &dummyMSAADProvider{
+		testUsers: buildTestUsers(),
+	}
+
+	msaad.SetProvider(testProvider)
+
+	sessionDB = newDummySessionDB()
+	permitDB = newDummyPermitDB()
+
+	groupIds := GroupIDU32s{GroupIDU32(0), GroupIDU32(1), GroupIDU32(2)}
+	var p Permit
+	p.Roles = EncodePermit(groupIds)
+	permitDB.SetPermit(1, &p)
+	p.Roles = EncodePermit(groupIds)
+	permitDB.SetPermit(2, &p)
+
+	roleDB = newDummyRoleGroupDB()
+	if err := roleDB.InsertGroup(&AuthGroup{
```

Comment:
Right! Just shove these function calls into this method! Why did I not think about that. No need for more methods, just have it all in one place. Very nice :)